src/amo/components/SearchPage/index.js

```diff
@@ -5,39 +5,89 @@ import { compose } from 'redux';
 
 import Search from 'amo/components/Search';
 import {
+  ADDON_TYPE_DICT,
+  ADDON_TYPE_EXTENSION,
+  ADDON_TYPE_LANG,
+  ADDON_TYPE_OPENSEARCH,
+} from 'core/constants';
+import { sendServerRedirect } from 'core/reducers/redirectTo';
+import {
   convertFiltersToQueryParams,
   convertQueryParamsToFilters,
 } from 'core/searchUtils';
 import type { ReactRouterLocation } from 'core/types/router';
 
 
 type Props = {|
+  clientApp: string,
+  dispatch: Function,
   filters: Object,
+  lang: string,
   location: ReactRouterLocation,
   pathname: string,
 |};
 
-export const SearchPageBase = ({ filters, pathname, ...props }: Props) => {
-  const paginationQueryParams = convertFiltersToQueryParams({
-    addonType: filters.addonType,
-    featured: filters.featured,
-    operatingSystem: filters.operatingSystem,
-    page: filters.page,
-    query: filters.query,
-    sort: filters.sort,
-    tag: filters.tag,
-  });
+export class SearchPageBase extends React.Component<Props> {
+  componentWillMount() {
+    const { clientApp, filters, lang, location } = this.props;
+
+    // Map the old `atype` parameter to its corresponding `adddonType`.
+    // See: https://github.com/mozilla/addons-frontend/issues/3791.
+    const newFilters = { ...filters };
+    if (location.query.atype) {
+      switch (String(location.query.atype)) {
```

Comment:
This logic seems like it would be better placed in a utility function (maybe inside `searchUtils`).

Reply:
I like the idea but I am not sure if we will reuse the logic somewhere else. Since it is a "patch" for a legacy behavior and it triggers a server-side redirect, maybe having everything in the same file makes everything more obvious. That being said, I have no preference.